firedrake/utility_meshes.py

```diff
@@ -863,6 +864,116 @@ def UnitIcosahedralSphereMesh(refinement_level=0, degree=1, reorder=None,
                                  comm=comm)
 
 
+def OctahedralSphereMesh(radius, refinement_level=0, degree=1,
+                         hemisphere=False, reorder=None,
+                         comm=COMM_WORLD):
+    """Generate an octahedral approximation to the surface of the
+    sphere.
+
+    :arg radius: The radius of the sphere to approximate.
+    :kwarg refinement_level: optional number of refinements (0 is an
+        octahedron).
+    :kwarg degree: polynomial degree of coordinate space (defaults
+        to 1: flat triangles)
+    :kwarg hemisphere: if True, mesh Northern Hemisphere only
+    :kwarg reorder: (optional), should the mesh be reordered?
+    :kwarg comm: Optional communicator to build the mesh on (defaults to
+        COMM_WORLD).
+    """
+    if refinement_level < 0 or refinement_level % 1:
+            raise RuntimeError("Number of refinements must be a non-negative integer")
+
+    if degree < 1:
+        raise ValueError("Mesh coordinate degree must be at least 1")
+    # vertices of an octahedron of radius 1
+    if hemisphere:
+        vertices = np.array([[1.0, 0.0, 0.0],
+                             [0.0, 1.0, 0.0],
+                             [0.0, 0.0, 1.0],
+                             [-1.0, 0.0, 0.0],
+                             [0.0, -1.0, 0.0]])
+        # faces of the base octahedron
+        faces = np.array([[0, 1, 2],
+                          [0, 2, 4],
+                          [1, 2, 3],
+                          [2, 3, 4]], dtype=np.int32)
+    else:
+        vertices = np.array([[1.0, 0.0, 0.0],
+                             [0.0, 1.0, 0.0],
+                             [0.0, 0.0, 1.0],
+                             [-1.0, 0.0, 0.0],
+                             [0.0, -1.0, 0.0],
+                             [0.0, 0.0, -1.0]])
+        # faces of the base octahedron
+        faces = np.array([[0, 1, 2],
+                          [0, 1, 5],
+                          [0, 2, 4],
+                          [0, 4, 5],
+                          [1, 2, 3],
+                          [1, 3, 5],
+                          [2, 3, 4],
+                          [3, 4, 5]], dtype=np.int32)
+
+    plex = mesh._from_cell_list(2, faces, vertices, comm)
+    plex.setRefinementUniform(True)
+    for i in range(refinement_level):
+        plex = plex.refine()
+
+    # build the initial mesh
+    m = mesh.Mesh(plex, dim=3, reorder=reorder)
+    if degree > 1:
+        # use it to build a higher-order mesh
+        new_coords = function.Function(functionspace.VectorFunctionSpace(m, "CG", degree))
+        x, y, z = ufl.SpatialCoordinate(m)
+        new_coords.interpolate(ufl.as_vector([x, y, z]))
+        m = mesh.Mesh(new_coords)
+
+    # remap to a cone
+    x, y, z = ufl.SpatialCoordinate(m)
+    tol = 1.0e-8
```

Comment:
1) this is unsafe, there is no guarantee `radius` is O(1)
2) if radius is 1, 1e-8 feels like an unnecessarily large amount to 'distort' the mesh by.  I'd suggest either reducing this even further, or writing a proper check s.t. if rnew < tol, x0 = y0 = 0.